### PR TITLE
Migrate to pyproject.toml, ruff, add build CI

### LINKED ISSUE #209: Migrate from setup.cfg to pyproject.toml?
Not sure what the advantages are but if someone is up for it, we can do it.

### PR DESCRIPTION
**Pull request by dcherian**
Closes #209

Also move to ruff so we can get rid of flake8 which means we can delete setup.cfg

Hopefully this gets us into github's dependency tracking

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,3 +1,118 @@
+[project]
+name = "cf-xarray"
```
Comment (dcherian) [inline pyproject.toml:2]:
```suggestion
name = "cf_xarray"
```

File: pyproject.toml [hunk 2]
```diff
@@ -1,3 +1,118 @@
+[project]
+name = "cf_xarray"
+description = "A convenience wrapper for using CF attributes on xarray objects"
+readme = "README.rst"
+requires-python = ">=3.8"
+license = {file = "LICENSE"}
+keywords = ["xarray", "metadata", "CF conventions"]
+classifiers = [
+    "Development Status :: 6 - Mature",
```
Comment (dcherian) [inline pyproject.toml:9]:
```suggestion
    "Development Status :: 5 - Production/Stable",
```

**Comment by dcherian** [comment]
@ocefpaf could you review this please

File: MANIFEST.in [hunk 3]
```diff
@@ -3,6 +3,7 @@ exclude cf_xarray/_version.py
 
 prune ci
```
Comment (dcherian) [inline MANIFEST.in:4]:
```suggestion
graft cf_xarray/scripts
prune ci
```

File: pyproject.toml [hunk 4]
```diff
@@ -1,3 +1,115 @@
...
+    "setuptools>=45",
+    "wheel",
+    "setuptools_scm[toml]>=6.2",
+    "setuptools_scm_git_archive",
+]
+build-backend = "setuptools.build_meta"
+
+[tool.setuptools]
+include-package-data = false
+packages = ["cf_xarray"]
```
Comment (dcherian) [inline pyproject.toml:41]:
```suggestion
packages = ["cf_xarray", "cf_xarray.scripts"]
```

File: pyproject.toml [hunk 5]
```diff
@@ -1,3 +1,115 @@
+[project]
+name = "cf_xarray"
+description = "A convenience wrapper for using CF attributes on xarray objects"
+readme = "README.rst"
+requires-python = ">=3.8"
+license = {file = "LICENSE"}
+keywords = ["xarray", "metadata", "CF conventions"]
+classifiers = [
+    "Development Status :: 5 - Production/Stable",
+    "License :: OSI Approved :: Apache License",
```
Comment (dcherian) [inline pyproject.toml:10]:
```suggestion
    "License :: OSI Approved :: Apache Software License",
```

File: pyproject.toml [hunk 6]
```diff
@@ -1,3 +1,115 @@
...
+
+[project.urls]
+homepage = "https://cf-xarray.readthedocs.io"
+documentation = "https://cf-xarray.readthedocs.io"
+repository = "https://github.com/xarray-contrib/cf-xarray.git"
+changelog = "https://github.com/xarray-contrib/cf-xarray/blob/main/doc/whats-new.rst"
+
+[build-system]
+requires = [
+    "xarray",
```
Comment (dcherian) [inline pyproject.toml:31]: Xarray is required for build so it can pull the version dynamically

Comment (ocefpaf) [reply pyproject.toml:31]: I would add that comment in that line to avoid future accidental removal of it.

File: pyproject.toml [hunk 7]
```diff
@@ -1,3 +1,115 @@
...
+    "setuptools>=45",
+    "wheel",
+    "setuptools_scm[toml]>=6.2",
+    "setuptools_scm_git_archive",
```
Comment (ocefpaf) [inline pyproject.toml:35]: If you are writing the version to a file I'm almost certain that you don't need this.

**Review by ocefpaf — APPROVED** [review]
LGTM. Just some minor comments.

File: .github/workflows/pypi.yaml [hunk 8]
```diff
@@ -1,26 +1,97 @@
...
+          python -m pip install --upgrade pip
+          python -m pip install dist/cf_xarray*.whl
+          python -m cf_xarray.scripts.print_versions
+
+      - name: Publish package to TestPyPI
+        if: github.event_name == 'push'
+        uses: pypa/gh-action-pypi-publish@v1.6.4
+        with:
+          user: __token__
+          password: ${{ secrets.TESTPYPI_TOKEN }}
```
Comment (dcherian) [inline .github/workflows/pypi.yaml:78]:
```suggestion
          password: ${{ secrets.TEST_PYPI_TOKEN }}
```

**Comment by dcherian** [comment]
Uploading to TestPyPI doesn't work 
1. Need to muck with version number
2. Needs a branch on the main repo to access the repo secret.

Let's deal with this later

**Pull request merged**